crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,240 @@
+        file_last_modified: SystemTime,
+    ) -> Option<FileCache> {
+        let files = self.files.lock().unwrap();
+        let file = files.get(path)?;
+
+        // Make sure the file hasn't changed since the cached run.
+        if file.last_modified != file_last_modified {
+            return None;
         }
 
-        s.end()
+        Some(file.clone())
+    }
+
+    /// Add or update a file cache at `path` relative to the package root.
+    pub(crate) fn update(&self, path: RelativePathBuf, file: FileCache) {
+        self.files.lock().unwrap().insert(path, file);
     }
-}
 
-struct SerializeMessage<'a> {
-    message: &'a Message,
-    file_id: usize,
+    /// Remove a file cache at `path` relative to the package root.
+    pub(crate) fn remove(&self, path: &RelativePath) {
+        self.files.lock().unwrap().remove(path);
+    }
 }
 
-impl Serialize for SerializeMessage<'_> {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let Message {
-            kind,
-            range,
-            fix,
-            // Serialized manually for all files
-            file: _,
-            noqa_offset: noqa_row,
-        } = self.message;
+/// [`Path`] that is relative to the package root in [`PackageCache`].
+pub(crate) type RelativePath = Path;
+/// [`PathBuf`] that is relative to the package root in [`PackageCache`].
+pub(crate) type RelativePathBuf = PathBuf;
 
-        let mut s = serializer.serialize_struct("Message", 5)?;
+/// On disk representation of the cache per source file.
+#[derive(Clone, Deserialize, Debug, Serialize)]
+pub(crate) struct FileCache {
+    /// Timestamp when the file was last modified before the (cached) check.
+    last_modified: SystemTime,
+    /// Imports made.
+    imports: ImportMap,
+    /// Diagnostic messages.
+    messages: Vec<CacheMessage>,
+    /// Source code of the file.
+    ///
+    /// # Notes
+    ///
+    /// This will be empty if `messages` is empty.
+    source: String,
+}
+
+impl FileCache {
+    /// Create a new source file cache.
+    pub(crate) fn new(
+        last_modified: SystemTime,
+        messages: &[Message],
+        imports: &ImportMap,
+    ) -> FileCache {
+        let source = if let Some(msg) = messages.first() {
+            msg.file.source_text().to_owned()
+        } else {
+            String::new() // No messages, no need to keep the source!
+        };
 
-        s.serialize_field("kind", &kind)?;
-        s.serialize_field("range", &range)?;
-        s.serialize_field("fix", &fix)?;
-        s.serialize_field("file_id", &self.file_id)?;
-        s.serialize_field("noqa_row", &noqa_row)?;
+        let messages = messages
+            .iter()
+            .map(|msg| CacheMessage {
+                kind: msg.kind.clone(),
+                range: msg.range,
+                fix: msg.fix.clone(),
+                noqa_offset: msg.noqa_offset,
+            })
+            .collect();
```

Comment:
I used custom serializers because cloning `kind` can be expensive (it contains up to three Strings). But it seems that this isn't impacting performance much?

Reply:
Because the cache is now shared between threads (also see https://github.com/astral-sh/ruff/pull/5120) we really need ownership of the values, dealing with the lifetime is quite painful. We could try using `Arc` or similar, but indeed cloning doesn't seem to impact performance that much.

Reply:
I guess one way to solve this would be to store a serialized message internally (so that it is only necessary to truncate all the file cache entries when serializing the overall cache) and deserialize the file cache when retrieving the content for a file. This would have the benefit that the cache remains lazy for most parts. But I can see how this complicates things and may not be worth the effort.